File: backend/app/schemas/student_opportunity.py
"""Pydantic schemas for the STUDENT side of opportunity discovery and
applications.

This is a thin, read-only *adapter* over the two real posting tables --
`internships` (database/migrations/018_internships.sql) and `jobs`
(019_jobs.sql). There is no `opportunities` table in this architecture and
none is introduced: the student-facing "opportunity" is just a normalized
view of one internship OR one job.

The student-facing identifier is a prefixed string --
``internship_<uuid>`` / ``job_<uuid>`` -- so the adapter can always
resolve which source table (and therefore which of
`applications.internship_id` / `applications.job_id`) a given opportunity
maps to, without assuming the raw internship/job UUIDs never collide.

An application row is still the existing `applications` row
(020_applications.sql) unchanged: `student_id` from the authenticated
caller, `industry_id` from the BEFORE-INSERT trigger, `opportunity_type` +
`internship_id`/`job_id` derived here, `status` defaulting to APPLIED.
The student never supplies any of those -- the only field they control is
`cover_note`.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# Mirrors applications.opportunity_type / the INTERNSHIP-xor-JOB split.
SourceType = Literal["INTERNSHIP", "JOB"]

# database/migrations/020_applications.sql -- applications.status CHECK.
# The student frontend renders all seven; the values are never redefined.
StudentApplicationStatus = Literal[
    "APPLIED",
    "UNDER_REVIEW",
    "SHORTLISTED",
    "INTERVIEW_SCHEDULED",
    "SELECTED",
    "REJECTED",
    "WITHDRAWN",
]


class OpportunityIndustry(BaseModel):
    """Company display info for a posting, read from `industry_profiles`
    (RLS: "Authenticated users can view industry profiles"). `company_name`
    can be null -- an Industry account can post before filling in its
    company profile -- and no `profiles` column (full_name, email, ...) is
    ever exposed here; there is no RLS path to it for a student."""

    id: str
    company_name: str | None = None
    industry_sector: str | None = None
    logo_url: str | None = None


class OpportunitySkill(BaseModel):
    skill_id: str
    skill_name: str
    category_name: str | None = None
    required_level: str
    importance: str


class StudentOpportunitySummary(BaseModel):
    """One published internship/job, normalized for the browse list."""

    id: str
    source_type: SourceType
    title: str
    description: str
    location: str | None = None
    work_mode: str | None = None
    status: str  # the posting's lifecycle status -- always PUBLISHED in the list
    industry: OpportunityIndustry | None = None
    application_deadline: str | None = None
    created_at: str | None = None
    has_applied: bool = False


class StudentOpportunityDetail(StudentOpportunitySummary):
    """One published internship/job, normalized for the detail page --
    every field either table can supply, plus the required-skill list.
    Fields the other source type doesn't have are simply null."""

    eligibility_criteria: str | None = None
    openings: int | None = None

    # internship-only
    duration_months: int | None = None
    stipend_amount: float | None = None
    stipend_currency: str | None = None
    start_date: str | None = None

    # job-only
    employment_type: str | None = None
    salary_min: float | None = None
    salary_max: float | None = None
    salary_currency: str | None = None
    experience_min_years: float | None = None

    skills: list[OpportunitySkill] = Field(default_factory=list)


class StudentOpportunityListResponse(BaseModel):
    opportunities: list[StudentOpportunitySummary]


class ApplyRequest(BaseModel):
    """POST /api/v1/student/opportunities/{id}/applications body.

    `cover_note` is the ONLY field a student may send. `extra="forbid"`
    structurally rejects any attempt to smuggle in `student_id`,
    `industry_id`, `status`, `match_score`, `internship_id`, `job_id`, or
    `opportunity_type` -- every one of those is derived server-side.
    """

    model_config = ConfigDict(extra="forbid")

    cover_note: str | None = Field(default=None, max_length=5_000)


class StudentApplicationOpportunity(BaseModel):
    """The posting an application points at, embedded on the student's own
    application list so "My Applications" needs no second round-trip. Can
    be partially null if the posting is no longer PUBLISHED (a student has
    no RLS path to a CLOSED/ARCHIVED posting)."""

    id: str
    source_type: SourceType
    title: str | None = None
    industry: OpportunityIndustry | None = None
    location: str | None = None


class StudentApplicationResponse(BaseModel):
    """A row from the existing `applications` table, student's own only.
    Same columns as the Industry response, minus nothing -- the student
    already owns every field here."""

    id: str
    student_id: str
    opportunity_type: str
    internship_id: str | None = None
    job_id: str | None = None
    status: str
    cover_note: str | None = None
    match_score: float | None = None
    applied_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    opportunity: StudentApplicationOpportunity | None = None


class StudentApplicationListResponse(BaseModel):
    applications: list[StudentApplicationResponse]


# ---- match (advisory, deterministic, reuses app.services.match_service) ----

SkillMatchStatus = Literal["MATCHED", "NEEDS_IMPROVEMENT", "MISSING"]
MatchRecommendation = Literal["STRONG", "GOOD", "PARTIAL", "LOW"]


class MatchSkill(BaseModel):
    skill_id: str
    skill_name: str
    required_level: str
    importance: str
    candidate_has: bool
    candidate_level: str | None = None
    candidate_verified: bool = False
    status: SkillMatchStatus


class OpportunityMatchResponse(BaseModel):
    """The authenticated student's own advisory skill fit against one
    opportunity. Computed fresh on every call by
    app.services.match_service.compute_match (the same deterministic
    engine the Industry applicant-match endpoint uses) -- no LLM, nothing
    stored, never coupled to an application row or its status."""

    opportunity_id: str
    score: int  # 0-100
    recommendation: MatchRecommendation
    skill_coverage: str
    required_count: int
    matched_count: int
    needs_improvement_count: int
    missing_count: int
    matched_skills: list[MatchSkill]
    needs_improvement_skills: list[MatchSkill]
    missing_skills: list[MatchSkill]

File: frontend/types/student-opportunity.ts
/**
 * Mirrors backend/app/schemas/student_opportunity.py -- field-for-field,
 * same nullability.
 *
 * IMPORTANT: this is a read adapter over the live `internships` / `jobs`
 * tables (database/migrations/018, 019). There is no `opportunities`
 * table and no `opportunity_id` column on `applications`. The
 * student-facing opportunity `id` is a prefixed string --
 * `internship_<uuid>` / `job_<uuid>` -- so the backend can always resolve
 * which source table a given opportunity maps to.
 *
 * Application `status` is one of the seven live `applications.status`
 * values (020_applications.sql) -- never a reduced set. The Industry
 * recruitment pipeline is the sole writer of it.
 */

import type { InterviewMode, InterviewStatus } from "@/types/interview";

export type SourceType = "INTERNSHIP" | "JOB";

/** Browse-list sort keys — a fixed whitelist mirroring backend
 * `OpportunitySort` (schemas/student_opportunity.py). `newest` is the
 * default and is never serialized to the URL or the API. */
export type OpportunitySort = "newest" | "deadline";

/** All seven live values -- see database/migrations/020_applications.sql. */
export type StudentApplicationStatus =
  | "APPLIED"
  | "UNDER_REVIEW"
  | "SHORTLISTED"
  | "INTERVIEW_SCHEDULED"
  | "SELECTED"
  | "REJECTED"
  | "WITHDRAWN";

export const STUDENT_APPLICATION_STATUSES: StudentApplicationStatus[] = [
  "APPLIED",
  "UNDER_REVIEW",
  "SHORTLISTED",
  "INTERVIEW_SCHEDULED",
  "SELECTED",
  "REJECTED",
  "WITHDRAWN",
];

/** Active candidate stages a student may still withdraw from. Mirrors the
 * backend `WITHDRAWABLE_STATUSES` (student_opportunity_service.py).
 * SELECTED (offer in hand), REJECTED (terminal) and WITHDRAWN (already
 * withdrawn) are not withdrawable — the backend returns 409 for those. */
export const WITHDRAWABLE_APPLICATION_STATUSES: readonly StudentApplicationStatus[] = [
  "APPLIED",
  "UNDER_REVIEW",
  "SHORTLISTED",
  "INTERVIEW_SCHEDULED",
];

export function canWithdrawApplication(status: StudentApplicationStatus): boolean {
  return WITHDRAWABLE_APPLICATION_STATUSES.includes(status);
}

export interface OpportunityIndustry {
  id: string;
  company_name: string | null;
  industry_sector: string | null;
  logo_url: string | null;
}

export interface OpportunitySkill {
  skill_id: string;
  skill_name: string;
  category_name: string | null;
  required_level: string;
  importance: string;
}

export interface StudentOpportunitySummary {
  id: string;
  source_type: SourceType;
  title: string;
  description: string;
  location: string | null;
  work_mode: string | null;
  status: string;
  industry: OpportunityIndustry | null;
  application_deadline: string | null;
  created_at: string | null;
  has_applied: boolean;
}

export interface StudentOpportunityDetail extends StudentOpportunitySummary {
  eligibility_criteria: string | null;
  openings: number | null;
  // internship-only
  duration_months: number | null;
  stipend_amount: number | null;
  stipend_currency: string | null;
  start_date: string | null;
  // job-only
  employment_type: string | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  experience_min_years: number | null;
  skills: OpportunitySkill[];
}

export interface StudentApplicationOpportunity {
  id: string;
  source_type: SourceType;
  title: string | null;
  industry: OpportunityIndustry | null;
  location: string | null;
  /** ONSITE / REMOTE / HYBRID / null (null once the posting is no longer
   * PUBLISHED). Drives the Applications-page workspace CTA for a SELECTED
   * internship: only REMOTE/HYBRID get an Internship Workspace. */
  work_mode: string | null;
}

/** The student's own LIVE (SCHEDULED) interview for an application.
 * Mirrors backend `StudentApplicationInterview`
 * (backend/app/schemas/student_opportunity.py), stitched on server-side
 * from the `public.student_interviews` RPC. `notes` is Industry-private
 * and is never included. `null` on `StudentApplication.interview` when the
 * application has no live interview (INTERVIEW_SCHEDULED can exist without
 * one — a manual status move, or a cancelled/completed interview). */
export interface StudentApplicationInterview {
  id: string;
  application_id: string;
  /** ISO-8601 UTC instant. Render in the viewer's locale. */
  scheduled_at: string;
  duration_minutes: number;
  mode: InterviewMode;
  location: string | null;
  status: InterviewStatus;
}

export interface StudentApplication {
  id: string;
  student_id: string;
  opportunity_type: SourceType;
  internship_id: string | null;
  job_id: string | null;
  status: StudentApplicationStatus;
  cover_note: string | null;
  match_score: number | null;
  applied_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  opportunity: StudentApplicationOpportunity | null;
  interview: StudentApplicationInterview | null;
}

export type SkillMatchStatus = "MATCHED" | "NEEDS_IMPROVEMENT" | "MISSING";
export type MatchRecommendation = "STRONG" | "GOOD" | "PARTIAL" | "LOW";

export interface MatchSkill {
  skill_id: string;
  skill_name: string;
  required_level: string;
  importance: string;
  candidate_has: boolean;
  candidate_level: string | null;
  candidate_verified: boolean;
  status: SkillMatchStatus;
}

export interface OpportunityMatch {
  opportunity_id: string;
  score: number;
  recommendation: MatchRecommendation;
  skill_coverage: string;
  required_count: number;
  matched_count: number;
  needs_improvement_count: number;
  missing_count: number;
  matched_skills: MatchSkill[];
  needs_improvement_skills: MatchSkill[];
  missing_skills: MatchSkill[];
}

/** Friendly label for every one of the seven statuses -- the backend
 * value is never changed, only presented. */
export const STATUS_LABEL: Record<StudentApplicationStatus, string> = {
  APPLIED: "Applied",
  UNDER_REVIEW: "Under Review",
  SHORTLISTED: "Shortlisted",
  INTERVIEW_SCHEDULED: "Interview Scheduled",
  SELECTED: "Selected",
  REJECTED: "Rejected",
  WITHDRAWN: "Withdrawn",
};
